docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+ * Runtime safety: A program that compiles is a program that runs safely, without crashes, traps or unhandled exceptions, including asynchronous behaviour, but not necessarily producing the right result or doing what you expect.
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can convert the object to a string representation
+interface Serialisable
+{
+  // All derived classes will have the common field
+  var common : F64;
+
+  // Derived classes must implement this method
+  toString() : String;
+};
+
+class MyType : Serialisable
+{
+  // Inherits `common`
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `toString`
+  toString() : String
+  {
+    ...
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
+
+They can also be a private member, if its name starts with `_`.
+
+They can also have a keyword `embed`, which means ...
+
+### Methods
+
+Methods are functions that belong to a class or interface. Like functions, they can have arguments and a return type.
+
+Methods can be called statically (via `Class::method()`) or dynamically (via `object.method()`).
+On the static case, the `self` pointer is a descriptor forthe type, while on the dynamic case, it's a descriptor to the instance (and its vtable, type info, etc).
+
+Static calls cannot change the fields of the class unless an object is being created and returned, like in the case of constructors.
+
+### Constructors and Destructors
+
+Verona distinguishes memory allocation with object construction.
+There are no `constructors` per se, in Verona, only static methods that can allocate memory (using `new`), build an object and return it.
+
+Example:
+```ts
+class Capibara
+{
+  let name : String;
+  var weight : F32; // Kgs
+
+  // Create a baby capibara
+  newborn(name : String) : Capibara
+  {
+    var self : Capibara = new Capibara;
+    self.name = name;
+    self.weight = 3; // Tiny baby!
+    self; // Returns `self`
+  }
+
+  eat(self: Capibara, food : F32)
+  {
+    self.weight += food;
+  }
+
+  poop(self: Capibara)
+  {
+    self.weight -= 0.1; // Tiny poop!
+  }
+};
+```
+
+On the example above, the `newborn` method creates a `Capibara` with a specific name and a default weight.
+The `name` can't change, so we declare it with `let`, while the `weight` does change, so it's a `var`.
```

Comment:
This is misleading. unless the String for name is immutable, it can still be changed with this `let` binding. The `let` is about the object the variable is bound to, not the mutability or lack thereof,

Reply:
Ah, yes, David has corrected with `& imm`, thanks!